From 289338fe3917d3ab25099c67c49035cf6381fc1e Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Thu, 25 Sep 2025 13:53:27 +0200
Subject: [PATCH 35/35] board: solidrun: lx2160acex7: change fan speed for
 lx2160 silicon r1
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

LX2160A silicon revision 1 has errata  when operating at junction
temperature beyond 85°C.

Configure fan-controller to run at full speed when reaching 85°C on
silicon 1. Later revisions have this problem resolved and can run stable
up to 105°C.

Reduce lower limit from 64°C to 60°C to ensure full speed is reached
before passing temperature limits.
Previous calculation was wrong in that full speed was reached only at
106.5°C which is above SoC maximum rating.

Finally remote one outdated function delcaration, fix 2 compiler
warnings for undeclared functions - and avoid using printing single %
character with printf (changed to puts).

Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 board/solidrun/lx2160acex7/lx2160a.c | 20 ++++++++++++++++----
 1 file changed, 16 insertions(+), 4 deletions(-)

diff --git a/board/solidrun/lx2160acex7/lx2160a.c b/board/solidrun/lx2160acex7/lx2160a.c
index ba2567d8774..478742ad587 100644
--- a/board/solidrun/lx2160acex7/lx2160a.c
+++ b/board/solidrun/lx2160acex7/lx2160a.c
@@ -19,6 +19,7 @@
 #include <init.h>
 #include <malloc.h>
 #include <i2c.h>
+#include <wdt.h>
 
 DECLARE_GLOBAL_DATA_PTR;
 
@@ -42,7 +43,6 @@ int board_early_init_f(void)
 }
 
 #ifdef CONFIG_OF_BOARD_FIXUP
-void board_fix_fdt_eth(void *fdt, u32 srds_s1, u32 srds_s2, u32 is_lx2162);
 void board_fix_fdt_serdes_ports(void *fdt);
 
 static void board_fix_fdt_pci_silicon(void *fdt) {
@@ -168,6 +168,7 @@ int board_init(void)
 static int setup_fan_ctrl(void) {
 	int ret = -ENODEV;
 	struct udevice *bus, *dev;
+	const char *config_str;
 
 	struct {
 		const char *const machine;
@@ -201,8 +202,17 @@ static int setup_fan_ctrl(void) {
 		if (ret)
 			continue;
 
-		/* set low temperatur threshold 64C, slope 1.57%/C, full-speed at 101C (safe for LX2160/LX2162 SoC) */
-		ret = dm_i2c_reg_write(dev, 0x25, 0x83);
+		/* set low temperature threshold and slope */
+		if (IS_SVR_REV(get_svr(), 1, 0)) {
+			/* LX2160A revision 1 is stable only up to 85°C Tj (errata A-011553) */
+			config_str = "Fan:   Low 60°C, High 81.25°C, Slope 3.14%\n";
+			ret = dm_i2c_reg_write(dev, 0x25, 0x7a);
+		} else {
+			/* LX2160A revision 2 / LX2162A revision 1 are stable up to 105°C Tj */
+			config_str = "Fan:   Low 60°C, High 102.5°C, Slope 1.57%\n";
+			ret = dm_i2c_reg_write(dev, 0x25, 0x7b);
+		}
+
 		if (ret)
 			continue;
 
@@ -210,7 +220,7 @@ static int setup_fan_ctrl(void) {
 		if (fanctrl[i].gpio_reg)
 			*fanctrl[i].gpio_reg &= ~fanctrl[i].gpio_mask;
 
-		printf("Fan:   Low 64°C, High 101°C, Slope 1.57%\n");
+		puts(config_str);
 	}
 
 	return ret;
@@ -260,6 +270,8 @@ void board_quiesce_devices(void)
 #endif
 
 #ifdef CONFIG_OF_BOARD_SETUP
+void board_setup_fdt_serdes_ports(void *fdt);
+
 int ft_board_setup(void *blob, struct bd_info *bd)
 {
 	int i;
-- 
2.51.0

